pypeit/edgetrace.py

```diff
@@ -960,384 +976,486 @@ def initial_trace(self, bpm=None, save=False):
-                                          name='CENTER_ERR'),
-                            fits.ImageHDU(header=mskhdr, data=self.spat_msk, name='CENTER_MASK'),
-                            fits.ImageHDU(header=fithdr, data=self.spat_fit.astype(float_dtype),
-                                          name='CENTER_FIT')])
-        # Add detector
-        hdu += self.trace_img.detector.to_hdu()
-        # Add order ID, if available
-        if self.orderid is not None:
-            hdu += [fits.ImageHDU(data=self.orderid, name='ORDERID')]
-        # Add PCA, if available
-        if self.pca is not None:
-            if self.par['left_right_pca']:
-                hdu += [self.pca[0].to_hdu(name='LPCA'), self.pca[1].to_hdu(name='RPCA')]
-            else:
-                hdu += [self.pca.to_hdu()]
-        # TODO: These things are going to go in slittrace.SlitTraceSet
-        if self.design is not None:
-            hdu += [fits.BinTableHDU(header=designhdr, data=self.design, name='DESIGN')]
-        if self.objects is not None: 
-            hdu += [fits.BinTableHDU(data=self.objects, name='OBJECTS')]
-        # Write the fits file; note not everything is written. Some
-        # arrays are reconstruced by the load function.
-        hdu.writeto(outfile, overwrite=True, checksum=checksum)
-
-        # Compress the file if the output filename has a '.gz'
-        # extension; this is slow but still faster than if you have
-        # astropy.io.fits do it directly
-        if compress:
-            msgs.info('Compressing file to: {0}.gz'.format(outfile))
-            io.compress_file(outfile, overwrite=overwrite)
-
-    def load(self, validate=True, rebuild_pca=False):
+    def _bundle(self):
+        """
+        Bundle the object for writing using
+        :func:`~pypeit.datamodel.DataContainer.to_hdu`.
         """
-        Load and reinitialize the trace data.
 
-        Data is read from :attr:`master_file_path` and used to
-        overwrite any internal data. Specific comparisons of the
-        saved data are performed to ensure the file is consistent
-        with having been written by a consistent version of the code;
-        see :func:`_reinit`.
+        # TODO: This will need to be edited if/when astropy.table.Table
+        # objects with the slit matching results get added to the
+        # datamodel.
 
-        To load a full :class:`EdgeTraceSet` directly from a file,
-        use :func:`from_file`.
+        # All of the datamodel elements that can be written as a header
+        # are written to the SOBELSIG extension.
+        d = []
+        for key in self.keys():
+            if self[key] is None:
+                continue
+            if isinstance(self[key], DataContainer):
+                d += [{key: self[key]}]
+                continue
+            if self.datamodel[key]['otype'] == np.ndarray:
+                # TODO: We might want a general solution in
```

Comment:
Ah, there we are.  I've noted in C that `bool` are also cast as `uint8` so this might be what is at the heart of this all.  That is, there is nothing less than `uint8` for I/O.